plugins/analogix/fu-analogix-device.c

```diff
@@ -0,0 +1,503 @@
+}
+
+static gboolean
+fu_analogix_device_send (FuAnalogixDevice *self,
+			 AnxBbRqtCode reqcode,
+			 guint16 val0code,
+			 guint16 index,
+			 const guint8 *buf,
+			 gsize bufsz,
+			 GError **error)
+{
+	gsize actual_len = 0;
+	g_autofree guint8 *buf_tmp = NULL;
+
+	g_return_val_if_fail (buf != NULL, FALSE);
+	g_return_val_if_fail (bufsz <= 64, FALSE);
+
+	/* make mutable */
+	buf_tmp = fu_memdup_safe (buf, bufsz, error);
+	if (buf_tmp == NULL)
+		return FALSE;
+
+	/* send data to device */
+	if (!g_usb_device_control_transfer (fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    reqcode, /* request */
+					    val0code, /* value */
+					    index, /* index */
+					    buf_tmp, /* data */
+					    bufsz, /* length */
+					    &actual_len, /* actual length */
+					    (guint) ANX_BB_TRANSACTION_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "send data error: ");
+		return FALSE;
+	}
+	if (actual_len != bufsz) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "send data length is incorrect");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_analogix_device_receive (FuAnalogixDevice *self,
+			    AnxBbRqtCode reqcode,
+			    guint16 val0code,
+			    guint16 index,
+			    guint8 *buf,
+			    gsize bufsz,
+			    GError **error)
+{
+	gsize actual_len = 0;
+
+	g_return_val_if_fail (buf != NULL, FALSE);
+	g_return_val_if_fail (bufsz <= 64, FALSE);
+
+	/* get data from device */
+	if (!g_usb_device_control_transfer (fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+					    G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    reqcode, /* request */
+					    val0code, /* value */
+					    index,
+					    buf, /* data */
+					    bufsz,  /* length */
+					    &actual_len, /* actual length */
+					    (guint) ANX_BB_TRANSACTION_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "receive data error: ");
+		return FALSE;
+	}
+	if (actual_len != bufsz) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "receive data length is incorrect");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_analogix_device_check_update_status (FuAnalogixDevice *self, GError **error)
+{
+	guint8 status = UPDATE_STATUS_INVALID;
+
+	for (guint i = 0; i < 3000; i++) {
+		if (!fu_analogix_device_receive (self,
+						 ANX_BB_RQT_GET_UPDATE_STATUS,
+						 0, 0,
+						 &status, sizeof(status),
+						 error))
+			return FALSE;
+		if (status == UPDATE_STATUS_ERROR) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_NOT_FOUND,
+					     "update status was error");
+			return FALSE;
+		}
+
+		/* assumed success */
+		if (status != UPDATE_STATUS_INVALID)
```

Comment:
this seems wrong - there is a status for start and finish, won't you mark it prematurely done if you end up getting `*_START`?

Reply:
I've renamed it `fu_analogix_device_get_update_status` and now check the value in the caller.